core/lib/zksync_core/src/consensus/storage.rs

```diff
@@ -0,0 +1,423 @@
+                head.header.number.next(),
+                block.header.number
+            )
+            .into());
+        }
+
+        storage
+            .put_block(ctx, block, self.operator_address)
+            .await
+            .wrap("put_block()")
+    }
+}
+
+#[async_trait::async_trait]
+impl BlockStore for SignedBlockStore {
+    async fn head_block(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        let head = *self.head.borrow();
+        storage
+            .find_head_forward(ctx, head, self.operator_address)
+            .await
+            .wrap("find_head_forward()")
+    }
+
+    async fn first_block(&self, ctx: &ctx::Ctx) -> ctx::Result<validator::FinalBlock> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        Ok(storage
+            .fetch_block(ctx, self.genesis, self.operator_address)
+            .await
+            .wrap("fetch_block()")?
+            .context("Genesis miniblock not present in Postgres")?)
+    }
+
+    async fn last_contiguous_block_number(
+        &self,
+        ctx: &ctx::Ctx,
+    ) -> ctx::Result<validator::BlockNumber> {
+        Ok(self.head_block(ctx).await?.header.number)
+    }
+
+    async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage()")?;
+        storage
+            .fetch_block(ctx, number, self.operator_address)
+            .await
+            .wrap("fetch_block()")
+    }
+
+    async fn missing_block_numbers(
+        &self,
+        _ctx: &ctx::Ctx,
+        _range: ops::Range<validator::BlockNumber>,
+    ) -> ctx::Result<Vec<validator::BlockNumber>> {
+        Ok(vec![]) // The storage never has missing blocks by construction
+    }
+
+    fn subscribe_to_block_writes(&self) -> sync::watch::Receiver<validator::BlockNumber> {
+        self.head.subscribe()
+    }
+}
+
+#[async_trait::async_trait]
+impl ReplicaStateStore for SignedBlockStore {
+    async fn replica_state(&self, ctx: &ctx::Ctx) -> ctx::Result<Option<ReplicaState>> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage")?;
+        Ok(ctx
+            .wait(storage.0.consensus_dal().replica_state())
+            .await?
+            .context("replica_state()")?)
+    }
+
+    async fn put_replica_state(
+        &self,
+        ctx: &ctx::Ctx,
+        replica_state: &ReplicaState,
+    ) -> ctx::Result<()> {
+        let storage = &mut storage(ctx, &self.pool).await.wrap("storage")?;
+        Ok(ctx
+            .wait(storage.0.consensus_dal().put_replica_state(replica_state))
+            .await?
+            .context("put_replica_state()")?)
+    }
+}
+
+#[async_trait::async_trait]
+impl PayloadSource for SignedBlockStore {
+    async fn propose(
```

Comment:
Dumb question: Does the consensus code expect that this method can take an extended period of time to complete? I.e., could this block other logic from executing / completing?

Reply:
Good point. propose() may be blocking in case there is nothing to propose for a while(no txs in mempool for example). Still it shouldn't just give up proposing right away, but rather wait, timing out only if the consensus moves to another view already. With the current implementation it will block both replica and the leader actors from processing messages. Replica definitely shouldn't be blocked by that. Leader should be fine in most cases with being blocked, but we should analyze that more thoroughly.